tests/ComTests.cpp

```diff
@@ -3058,4 +3060,175 @@ TEST_CASE("COMEnumerator", "[com][enumerator]")
 
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP | WINAPI_PARTITION_SYSTEM)
+#if (NTDDI_VERSION >= NTDDI_WINBLUE)
+#if defined(__cpp_impl_coroutine) || defined(__cpp_coroutines) || defined(_RESUMABLE_FUNCTIONS_SUPPORTED)
+
+#include <wil/cppwinrt_register_com_server.h>
+#include <winrt/windows.foundation.h>
+#include <windows.foundation.h>
+
+const winrt::guid CLSID_RPCTimeoutTestServer{L"CED2F47C-200B-476F-BFDC-D0D79B052AC6"};
+HANDLE g_hangHandle{nullptr};
+HANDLE g_doneHangingHandle{nullptr};
+
+TEST_CASE("com_timeout", "[com][com_timeout]")
+{
+    auto init = wil::CoInitializeEx_failfast();
+
+    // These test cases require a COM server that will use RPC for communication so that we can exercise
+    // RPC cancellation.  Additionally, this server needs to support the ability to control if it hangs
+    // or returns quickly.  The following class provides that functionality, using some global events to
+    // provide deterministic ordering of operations.
+    //
+    // Normally a server in the same process would not use RPC.  By marking this as non_agile and running
+    // it on an STA thread we are able to get RPC involved enough to test cancellation.
+    struct RPCTimeoutTestServer : winrt::implements<RPCTimeoutTestServer, winrt::Windows::Foundation::IStringable, winrt::non_agile>
+    {
+        winrt::hstring ToString()
+        {
+            // If the global handle exists, block on it.  If it doesn't exist then this is a non-hang case
+            // so skip waiting.
+            if (g_hangHandle)
+            {
+                // Pump messages so this STA thread is healthy while we wait.  If this wait fails that means
+                // the cancel did not work.
+                HANDLE handles[1] = {g_hangHandle};
+                DWORD index;
+                REQUIRE_SUCCEEDED(CoWaitForMultipleObjects(
+                    CWMO_DISPATCH_CALLS | CWMO_DISPATCH_WINDOW_MESSAGES, 10000, ARRAYSIZE(handles), handles, &index));
+
+                if (g_doneHangingHandle)
+                {
+                    SetEvent(g_doneHangingHandle);
+                }
+            }
+            return L"RPCTimeoutTestServer";
+        }
+    };
+
+    // The COM server thread needs an event that is signaled when we want it to stop pumping messages and
+    // exit.
+    wil::shared_event comServerEvent;
+    comServerEvent.create();
+
+    auto comServerThread = std::thread([comServerEvent] {
+        // This thread must be STA to pull RPC in as mediator between threads.
+        auto init = wil::CoInitializeEx_failfast(COINIT_APARTMENTTHREADED);
+
+        DWORD registration{};
+        REQUIRE_SUCCEEDED(CoRegisterClassObject(
+            winrt::guid{CLSID_RPCTimeoutTestServer},
+            winrt::make<wil::details::CppWinRTClassFactory<RPCTimeoutTestServer>>().get(),
+            CLSCTX_LOCAL_SERVER,
+            REGCLS_MULTIPLEUSE,
+            &registration));

```

Comment:
Would arguably be more straightforward to use agile_ref to create the proxy, rather than activation:
* Add an agile_ref<IStringable> to the MTA scope, ref-captured by the lambda
* On the STA thread, use make_agile to wrap the object in the agile_ref. This replaces CoRegisterClassObject.
* On the MTA thread, use <agileRef>.get() to unwrap the agile_ref. The result is a proxy connected to the STA object. This replaces the CoCreateInstance.

Reply:
I was gonna say "that doesn't sound much simpler" but then I tried coding it up and... yeah its simpler.  It also saves me from having that extremely ugly CoCreate polling loop.